pycalphad/mapping/utils.py

```diff
@@ -0,0 +1,70 @@
+from typing import Union
+
+import numpy as np
+
+from pycalphad import variables as v
+from pycalphad.core.composition_set import CompositionSet
+
+from pycalphad.mapping.primitives import Point, Node, STATEVARS
+
+def degrees_of_freedom(point: Union[Point, Node], components: list[str], num_potential_conditions: int):
+    non_va_comps = len(set(components) - {"VA"})
+    return non_va_comps + 2 - len(point.stable_composition_sets) - (2 - num_potential_conditions)
+
+def update_cs_phase_frac(comp_set: CompositionSet, phase_frac: float):
+    comp_set.update(comp_set.dof[len(STATEVARS):], phase_frac, comp_set.dof[:len(STATEVARS)])
+
+def get_statevars_array(conditions: dict[v.StateVariable, float]):
+    return np.asarray([conditions[sv] for sv in STATEVARS], dtype=np.float64)
+
+def elements_from_components(components: list[str]):
```

Comment:
There used to be pycalphad code for getting nonvacant pure elements, might be good to use that instead if we can